### ISSUE TITLE
idea: leverage module augmentation to reduce bundle size

### ISSUE THREAD

**Issue by mfulton26** [colinhacks/zod#3623]
I like how Zod has great auto-completion, but it currently results in unused code being loaded which increases bundle sizes (e.g. see #2596).

[Module augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) could be a great way to reduce the core Zod size and allow developers to optionally load additional schema methods. Zod can still be a large library. In fact, it can grow as much as wanted with lots of methods. Consumers will still only load what they want. Their bundle sizes will remain small and only grow as they opt-into more methods. This results in faster loading web pages, cloud function (e.g. AWS Lambda), etc.

This could be a breaking change or Zod could continue to export everything together but provide alternative import paths for this module augmentation leveraging approach.

Here is an oversimplified example that only loads desired methods:

```ts
import { z } from "./zod";
import "./zod/string";
import "./zod/string/max";

console.log(z.string().max(12).parse("abc")); // => "abc"

// cannot use .number() unless it is loaded by importing ./zod/number"
// console.log(z.number().parse(42));
//               ^^^^^^ error TS2339: Property 'number' does not exist on type 'Zod'.

```

<details>
<summary>zod/index.ts</summary>

```ts
export class Zod {}

export const z = new Zod();

type ZodCheck<T> = (value: T) => void;

export abstract class ZodSchema<T> {
  #checks: ZodCheck<T>[];

  constructor(checks: ZodCheck<T>[] = []) {
    this.#checks = checks;
  }

  protected abstract typeCheck(value: unknown): asserts value is T;

  parse(value: unknown) {
    this.typeCheck(value);
    for (const check of this.#checks) check(value);
    return value;
  }

  refinement(check: ZodCheck<T>): this {
    type Self = this;
    return new (this.constructor as { new (checks: ZodCheck<T>[]): Self })([
      ...this.#checks,
      check,
    ]);
  }
}

```

</details>

<details>
<summary>zod/number.ts</summary>

```ts
import type { Zod } from ".";
import { z, ZodSchema } from ".";

export class ZodNumber extends ZodSchema<number> {
  protected typeCheck(value: unknown): asserts value is number {
    if (typeof value === "number") return;
    throw new Error("value is not a number");
  }
}

declare module "." {
  interface Zod {
    number(): ZodNumber;
  }
}

z.number = function (this: Zod) {
  return new ZodNumber();
};

```

</details>

<details>
<summary>zod/string.ts</summary>

```ts
import type { Zod } from ".";
import { z, ZodSchema } from ".";

export class ZodString extends ZodSchema<string> {
  protected typeCheck(value: unknown): asserts value is string {
    if (typeof value === "string") return;
    throw new Error("value is not a string");
  }
}

declare module "." {
  interface Zod {
    string(): ZodString;
  }
}

z.string = function (this: Zod) {
  return new ZodString();
};

```

</details>

<details>
<summary>zod/string/max.ts</summary>

```ts
import { ZodString } from "../string";

declare module "../string" {
  interface ZodString {
    max(maxLength: number): ZodString;
  }
}

Object.defineProperty(ZodString.prototype, "max", {
  value(this: ZodString, maxLength: number) {
    return this.refinement((value) => {
      if (value.length <= maxLength) return;
      throw new RangeError("max length exceeded");
    });
  },
});

```

</details>

**Comment by Christopher96u** [comment]
Hi, now zod v4 has reduced more than 50% its bundle size, but if you want something even smaller. There is zod/mini package with the smallest package size.

Please, have a look here [https://v4.zod.dev/packages/mini](url) and if this answer your question, could you please close the issue? Cheers

**Comment by mfulton26** [comment]
> Hi, now zod v4 has reduced more than 50% its bundle size, but if you want something even smaller. There is zod/mini package with the smallest package size.
> 
> Please, have a look here [https://v4.zod.dev/packages/mini](url) and if this answer your question, could you please close the issue? Cheers

That's neat. Tree-shaking is neat. For those not using a bundler that supports it though (e.g. loading things directly via JavaScript module imports from a CDN, Vite/Rollup, etc.) then tree-shaking doesn't help there.

I looked and from what I can tell all of `@zod/mini`'s exports are from the top-level entry-point. Can an entry-point be made for each function instead? This way developers would be able to import only what they want without any need for tree-shaking (which is mostly what I'm after, even if it isn't module augmentation based).

e.g. Instead of this

```ts
import * as z from "@zod/mini";
 
const mySchema = z.nullable(z.optional(z.string()));
```

I would like to be able to do something like this:

```ts
import nullable from "npm:@zod/mini/nullable";
import optional from "npm:@zod/mini/optional";
import string from "npm:@zod/mini/string";

const mySchema = nullable(optional(string()));
```

**Comment by mfulton26** [comment]
The bundle size reduction though is great news as is `@zod/mini`. Thank you!

**Comment by mfulton26** [comment]
Yes, this issue is still relevant to the latest version of Zod. I'd like to see individual entry points exported by Zod so that developers can load only what they want to keep overall bundle sizes low without needing to concern themselves with tree-shaking. I personally would like to see this with regular Zod via module augmentation as originally explained as it maintains a chainable schema builder API but this is also applicable to Zod Mini by allowing developers to only import the functions they want from smaller modules.

**Comment by mfulton26** [comment]
That's odd, right? I commented but the issue still got closed. 😞

**Comment by temrb** [comment]
had to revert to v3.. 

v4 (even mini) is like 4x larger & bloats project with unnecessary etc items.

**Comment by mfulton26** [comment]
Yes, this issue is still relevant. Thank you.

**Comment by colinhacks** [comment]
Your proposal is interesting, but I want a different way with it and added Zod Mini. I did consider this approach, but the idea of each user of Zod needing to add side-effectful imports for each method they wanted to use seemed a bit overwrought to me. https://zod.dev/packages/mini
